Add more analytics for hot reload in flutter_tools.

chingjun opened the pull request:
## New fields for existing analytics
### Hot reload
* `compileTimeInMs`: Time spent in the resident compiler to compile the changes
* `findInvalidatedTimeInMs`: Time spent in `projectFileInvalidator.findInvalidated`
* `scannedSourcesCount`: Number of source files scanned in `projectFileInvalidator.findInvalidated`
* `reassembleTimeInMs`: Time spent calling `device.vmService.flutterReassemble` or fast reassemble
* `reloadVMTimeInMs`: Time spent calling `device.vmService.service.reloadSources`
* Existing `transferTimeInMs` repurposed to be the actual time spent transferring data onto the device. Previously the stopwatch wasn't stopped so it was the same as `overallTimeInMs`.

### Hot restart (when success)
* `overallTimeInMs`: Total time spent during hot restart
* `syncedBytes`: Total data transferred onto the device
* `scannedSourcesCount`: Number of source files scanned in `projectFileInvalidator.findInvalidated`
* `invalidatedSourcesCount`: Number of source files invalidated
* `findInvalidatedTimeInMs`: Time spent in `projectFileInvalidator.findInvalidated
* `transferTimeInMs`: Time spent transferring data onto the device
* `compileTimeInMs`: Time spent in the resident compiler to compile the changes

## Newly added event
### When hot reload is first ready after flutter_tools is started
* `targetPlatform`: Self explanatory
* `sdkName`: Self explanatory
* `emulator`: Whether the app is running on an emulator
* `overallTimeInMs`: Time from `HotRunner.run()` is called to when hot reload is ready
* `compileTimeInMs`: Time spent initializing the resident compiler
* `transferTimeInMs`: Time spent launching the app on the device (i.e. time spent in `device.runHot`)

chingjun committed:
Add more analytics for hot reload in flutter_tools.

chingjun commented:
@jonahwilliams , I'm sending a draft PR here and would like to hear what you think about the new analytics added here, and the approach of collecting the data.

If it looks good to you I'll work on adding tests. Thanks :)

jonahwilliams reviewed (commented):
This looks great! Also of interest since I was scratching my head a bit trying to figure out compile time for @jensjoha

chingjun committed:
Add tests

chingjun commented:
Thanks Jonah, I've just added tests for this.

Sorry for the huge diff, here's a summary of what I did to add tests (other than the tests themselves):

1. Added StopwatchFactory to HotRunner, and moved StopwatchFactory from logger_test.dart to fakes.dart
2. Added a `name` parameter to `createStopwatch` to track the timers when a single test creates multiple stopwatches.
3. Slight refactor of HotRunner, moving some part of the code into `ReloadSourcesHelper` and `ReassembleHelper` so that the parts can be faked in tests.

PTAL, thanks!

jonahwilliams commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -119,6 +129,29 @@ class HotRunner extends ResidentRunner {
+  bool _emulator;
+
+  Future<void> _calculateTargetPlatform() async {
+    if (_targetPlatform == null) {
```
nit, maybe `if (_targetPlatform != null) return;` and then un-indent

chingjun replied:
Done.

jonahwilliams commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -1041,10 +1112,154 @@ class HotRunner extends ResidentRunner {
+}
+
+@visibleForTesting
+class ReassembleHelper {
```
Since this class doesn't have any state or dependencies, you could instead use a function: `typedef ReassembleHelper = Future<ReassembleResult> reassemble(...)` and then inject that

chingjun replied:
Done.

jonahwilliams commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -698,14 +765,16 @@ class HotRunner extends ResidentRunner {
-        reason: reason,
-        fastReassemble: null,
-      ).send();
+      if (restartEvent != null) {
```
this shouldn't ever be null right?

chingjun replied:
I added new fields when it succeed. In here, if hot restart succeeded, the event will be sent above, and restartEvent will be null. The code here is to handle failure cases only. Added comment to explain.

jonahwilliams replied:
Ahh, got it

jonahwilliams commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -342,14 +382,35 @@ class HotRunner extends ResidentRunner {
+      }));
     }
+
+    unawaited(appStartedCompleter?.future?.then((_) => HotEvent('reload-ready',
```
this is a new event right?

chingjun replied:
That's right. It's to track the time it spent to launch the app and get hot reload to be ready.

jonahwilliams replied:
instead of attaching it to this future, why not send it in where appStartedCompleter.complete is called? That would be at right about the same time

chingjun replied:
The only reason I do it here is because all the `Stopwatch`es  that collect the metrics are in this method, and `appStartedCompleter.complete` is called from another method. E.g. the `appStartedTimer`, is started when this method starts, and ended when `appStartedCompleter.complete` is called.

Or do you think it's better if I pass the `Stopwatch`, and two existing `Duration`s into the `attach()` method instead?

jonahwilliams replied:
no, I think changing the signature of attach is much worse. this does feel awkward though.

jonahwilliams commented on packages/flutter_tools/test/general.shard/hot_test.dart:
```diff
@@ -297,8 +445,25 @@ void main() {
 }
 
 class FakeDevFs extends Fake implements DevFS {
+  // UpdateFSReport updateReport;
```
remove

chingjun replied:
Oops, missed this one. Removed.

chingjun committed:
Updates based on review feedback

chingjun commented:
Updated, PTAL, thanks!

By the way, can you point me at how to `ensure that the custom dimension is defined in the backend, or will be defined shortly after the relevant PR lands` as instructed in custom dimensions?

jonahwilliams reviewed (approved):
LGTM
